inline.h

```diff
@@ -57,6 +57,94 @@ Perl_av_count(pTHX_ AV *av)
     return AvFILL(av) + 1;
 }
 
+/* ------------------------------- av.c ------------------------------- */
+
+/*
+=for apidoc av_store_simple
+
+This is a cut-down version of av_store that assumes that the array is
+very straightforward - no magic, not readonly, and AvREAL - and that
+C<key> is not negative. This function MUST NOT be used in situations
+where any of those assumptions may not hold.
+
+Stores an SV in an array.  The array index is specified as C<key>.  The
+return value will be C<NULL> if the operation failed or if the value did not
+need to be actually stored within the array (as in the case of tied
+arrays).  Otherwise, it can be dereferenced
+to get the C<SV*> that was stored
+there (= C<val>)).
+
+Note that the caller is responsible for suitably incrementing the reference
+count of C<val> before the call, and decrementing it if the function
+returned C<NULL>.
+
+Approximate Perl equivalent: C<splice(@myarray, $key, 1, $val)>.
+
+=cut
+*/
+
+PERL_STATIC_INLINE SV**
+Perl_av_store_simple(pTHX_ AV *av, SSize_t key, SV *val)
+{
+    SV** ary;
+
+    PERL_ARGS_ASSERT_AV_STORE_SIMPLE;
+    assert(SvTYPE(av) == SVt_PVAV);
+    assert(!SvMAGICAL(av));
+    assert(!SvREADONLY(av));
+    assert(AvREAL(av));
+    assert(key > -1);
+
+    ary = AvARRAY(av);
+
+    if (AvFILLp(av) < key) {
+        if (key > AvMAX(av)) {
+            av_extend(av,key);
+            ary = AvARRAY(av);
+        }
+        AvFILLp(av) = key;
+    } else
+        SvREFCNT_dec(ary[key]);
+
+    ary[key] = val;
+    return &ary[key];
+}
+
+/*
+=for apidoc av_fetch_simple
+
+This is a cut-down version of av_fetch that assumes that the array is
+very straightforward - no magic, not readonly, and AvREAL - and that
+C<key> is not negative. This function MUST NOT be used in situations
+where any of those assumptions may not hold.
+
+Returns the SV at the specified index in the array.  The C<key> is the
+index.  If lval is true, you are guaranteed to get a real SV back (in case
+it wasn't real before), which you can then modify.  Check that the return
+value is non-null before dereferencing it to a C<SV*>.
```

Comment:
I think this should also mention, similarly to store_simple, that refcount management is the caller's responsibility.

It might also be worth mentioning that one should not invoke any perl code (ie anything that could modify the av) between calling this and reading the SV**.

Reply:
Just thinking out loud: for both functions, given the cut-down nature, would there be any circumstances in which the caller would have to decrement the ref count?

The original `av_fetch` docs don't warn about not doing anything that could modify the av prior to reading the SV**. I guess that any added statement should be replicated there too. Is that a real concern for perl code though, rather than any core/XS users of `av_fetch`/`_simple`?

Reply:
> Just thinking out loud: for both functions, given the cut-down nature, would there be any circumstances in which the caller would have to decrement the ref count?

Not for errors, I think; however the caller may or may not need to increment.

> The original `av_fetch` docs don't warn about not doing anything that could modify the av prior to reading the SV**. I guess that any added statement should be replicated there too. Is that a real concern for perl code though, rather than any core/XS users of `av_fetch`/`_simple`?

My concern is that we don't explicitly say that the `SV**` returned is a pointer directly into the `AvARRAY`, so just based on the documentation this doesn't seem obvious to me. I'd consider it a safer function if it just returned an `SV*`.

Reply:
> I think this should also mention, similarly to store_simple, that refcount management is the caller's responsibility.

Looking at this again, does `av_fetch_simple` really need this? Note that `lval` here is basically a flag, not a SV*.



> My concern is that we don't explicitly say that the `SV**` returned is a pointer directly into the `AvARRAY`, so just based on the documentation this doesn't seem obvious to me. I'd consider it a safer function if it just returned an `SV*`.

This is what `av_fetch` does though and I wanted `av_fetch_simple` to be a drop-in replacement.

Reply:
> > I think this should also mention, similarly to store_simple, that refcount management is the caller's responsibility.
> 
> Looking at this again, does `av_fetch_simple` really need this?

Need? No. Would the docs be improved by such a change? I'm inclined towards yes, but I'm by no means certain - I do understand that too much doc can be as bad as too little.

> Note that `lval` here is basically a flag, not a SV*.

Yeah, that also took me a few beats to understand, and I had considered suggesting renaming it to `want_lval` (and perhaps re-typing it to bool). But if that's valuable it should be a separate change that also covers av_fetch.

> > My concern is that we don't explicitly say that the `SV**` returned is a pointer directly into the `AvARRAY`, so just based on the documentation this doesn't seem obvious to me. I'd consider it a safer function if it just returned an `SV*`.
> 
> This is what `av_fetch` does though and I wanted `av_fetch_simple` to be a drop-in replacement.

Yes, I'm not actually suggesting to change the return type. And again, if the doc is worth clarifying in this way it would be sensible to do it for av_fetch too.

As I mentioned at the start, I think it would be fine to consider all these doc changes separately after merging this. I'd be happy to put together a PR for that.

Reply:
I've not made any changes here.